src/plugin/svipc/sysvipcwrappers.cpp

```diff
@@ -183,9 +188,72 @@ shmdt(const void *shmaddr)
 // in order to install its own hooks. For us, shmdt() is the only interesting
 // one. Instead of giving the address of our wrapper to the hook library, we
 // want to return the address in libc. See PR #472 for details.
+static int dlsym_addr_instance = 0;
+static const char *dlsym_symbol = NULL;
+static void *dlsym_retval;
+// extern "C"
+static int
+callback(struct dl_phdr_info *info, size_t size, void *data) {
+  void *handle = dlopen(info->dlpi_name, RTLD_LAZY);
+  dlerror(); // initialize for errors
+  void *address = dlsym(handle, dlsym_symbol);
+  if (address == NULL && dlerror() != NULL) {
+    dlsym_retval = dlsym(handle, dlsym_symbol); // Set dlerror to error string
+    dlclose(handle);
+    return 0;
+  } else {
+    dlclose(handle);
+  }
+  if (address != NULL) {
+    dlsym_addr_instance++;
+    printf("Symbol %s in %s: found at address %p\n",
+           dlsym_symbol, info->dlpi_name, address);
+  }
+  if (dlsym_addr_instance == 1) {
+    printf("Caller of %s() in %s: wrapper fnc. address:  %p\n",
+           dlsym_symbol, info->dlpi_name, address);
+  }
+  if (dlsym_addr_instance == 2) {
+    printf("RTLD_NEXT:  address:  %p\n", address);
+    dlsym_retval = address;
+  }
+  return 0;
+}
+
+// extern "C"
+static void *
+dlsym_with_rtld_next(const char *symbol) {
+  dlsym_symbol = symbol;
+  dlsym_addr_instance = 0;
+  dlsym_retval = NULL;
+  dl_iterate_phdr(callback, NULL);
+  return dlsym_retval;
+}
+
+//   We need to wrap dlsym to disable and re-enable ckpt.  But then,
+// dlsym(RTLD_NEXT, ...) no longer works, since glibc:dlsym() looks up
+// the stack to find the caller's library, and discovers a DMTCP library
+// instead of the original target library.
+//   It was originally placed in the svipc plugin so that RTLD_NEXT would
+// point to glibc.  But some targets want RTLD_NEXT to point to a library
+// before glibc.  This happened when ompt_start_tool called dlsym()
+// to see if there was a preferred definition later in the search path.
+// By default, dlsym() should return NULL.  But now, dlsym is called by
+// DMTCP, and so dlsysm discovers the Intel OpenMP library (libiomp5.so)
+// later in the search path.
+// FIXME:  This code belongs in 'src' in libdmtcp.so, and not here.
+// FIXME:  This code searches from the beginning of the link map, using
```

Comment:
@gc00: I am not sure if this logic would work. Even some of our plugins use RTLD_NEXT and if the same symbol is wrapped multiple times, we'll be in trouble if we always skip the first match only. I think what we need is to record the return-address and use that to figure out the _next_ symbol.

Reply:
> ... we'll be in trouble if we always skip the first match only. I think what we need is to record the return-address and use that to figure out the next symbol.

I assume you meant to record the address of the symbol in the current library, and then figure out the next symbol.  But then, what if 'dlsym' is called in a library that doesn't define the target symbol?

I suppose we could add that heuristic on top of the current heuristic.  And we could even do things like save old values of dlsym, to decide if RTLD_NEXT is making progress or getting stuck at one library.  But before we go too deeply down that road, let's talk interactively.

For now, I think the perfect is the enemy of the good.  And I'll note that currently, this PR passes all of our tests.  I propose that talk about the roadmap for dlsym, and then decide what to commit now, and what to add a FIXME for.

In general, I don't see a simple fix for all cases.  And this is needed by one of the DMTCP "customers".  They define a function that is _not_ part of glibc.  And they use dlsym to test if their is a later definition of the same function.  If there is not, then they accept the definition of the current library as the "default".  I intend to add a test for this case.  We were failing before, for this case.

Reply:
Let's assume A->B->C->libdmtcp.so->D->libc.so, where, A-D are plugin libraries. If A, B and D had wrappers for a function F(), would this code return the correct value when calling dlsym(RTLD_NEXT, "F") from B or D?

For handling cases where C calls dlsym(RTLD_NEXT, "F"), we should have a thread-local variable `curLib` which is used to record info about the library that called dlsym(RTLD_NEXT, "F") (e.g., by lookup up the return address). Then we continue returning `0`, until we see curLib in the callback, i.e.:
```
callback(struct dl_phdr_info *info, size_t size, void *data) {
  if (curLib != NULL && curLib != info) {
    return 0;
  }

  if (curLib == info) {
    curLib = NULL;
    return  0;
  }
 ...
}
 ```